src/pymor/algorithms/qr.py

```diff
@@ -0,0 +1,67 @@
+# This file is part of the pyMOR project (https://www.pymor.org).
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+
+from pymor.algorithms.gram_schmidt import gram_schmidt
+
+
+def qr(A, product=None, offset=0, check=True, check_tol=1e-3, copy=True):
+    """Compute a QR decomposition.
+
+    Parameters
+    ----------
+    A
+        The |VectorArray| for which to compute the QR decomposition.
+    product
+        The inner product |Operator| w.r.t. which to orthonormalize.
+        If `None`, the Euclidean product is used.
+    offset
+        Assume that the first `offset` vectors are already orthonormal and start the
+        algorithm at the `offset + 1`-th vector.
+    check
+        If `True`, check if the resulting |VectorArray| is really orthonormal.
+    check_tol
+        Tolerance for the check.
+    copy
+        If `True`, create a copy of `A` instead of modifying `A` in-place.
+
+    Returns
+    -------
+    Q
+        The orthonormal |VectorArray| of the same length as `A`.
+    R
+        The upper-triangular/trapezoidal matrix.
+    """
+    return gram_schmidt(A, product=product, return_R=True, atol=0, rtol=0, offset=offset,
+                        check=check, check_tol=check_tol, copy=copy)
+
+
+def rrqr(A, product=None, offset=0, check=True, check_tol=1e-3, copy=True):
+    """Compute a rank-revealing QR (RRQR) decomposition.
+
```

Comment:
Document that rank-revealing means that `Q` may contain less vectors than `A`.

Reply:
Additionally, what would be the equivalent to the $A\approx QR$ reconstruction criterion?

Reply:
Not sure I follow. $A \approx QR$ should still hold, even in the rank-revealing version.

Reply:
My thought was that in this case we are more interested in getting minimum rank $Q$ such that the column spans of $A$ and $Q$ are the same. Technically that should be true when the approximation of $A$ is good enough, but I am wondering if something else like the subspace angle would be better. However, computing that via principal angles would involve orthogonalization in the first place...

Reply:
I feel like this complicates things too much. Also even for rrqr one might be interested in $\|A - QR\|$. So this would be yet a third criterion to check.